New user package for semi infinite crack propagation

bnathan2 opened the pull request:
**Description**

Attached is a version of LAMMPS with an added home-built user package that introduces and propagates a semi-infinite crack through a sample using the equations of linear elastic fracture mechanics in conjunction with molecular dynamics.  The package directory has been included in the /src file along with a README.txt file, a documentation file, .cpp and .h files and 2 sample input scripts.  Our hope is for this package to be made available as a standard user package for LAMMPS.  To this end, we invite all questions and comments from the core developers.

**Author(s)**

Nathan Beets, Graduate Research Assistant, Virginia Tech Department of Materials Science and Engineering. (corresponding author) 
email: nathanbeets324@gmail.com 

Dr. Diana Farkas, Professor, Virginia Tech Department of Materials Science and Engineering. 
email: diana@vt.edu

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

This code is backwards compatible. 

**Implementation Notes**

No other features in LAMMPS are affected by the inclusion of this package. 

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

_Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)_

akohlmey commented on bench/in.sicrackfoam_M1:
```diff
@@ -0,0 +1,62 @@
+#set		    group boundary type 2
+
+pair_style eam
+pair_coeff * * /Users/bnathan2/Desktop/LAMMPS_git_Aug14_2019/potentials/Au_u3.eam
```
please don't put absolute paths in input examples, and definitely not paths that won't work. also the data file is WAAAY too large for inclusion in the distribution and it should not be added into the bench folder. please also put a log command into example inputs.

akohlmey commented on src/USER-SICRACK/Install.csh:
```diff
@@ -0,0 +1,13 @@
+# Install/unInstall package "USER-SICRACK" classes in LAMMPS
```
we use bash script only. no c-shell.

akohlmey commented on src/USER-SICRACK/Package_SiCrack_Documentation.txt:
```diff
@@ -0,0 +1 @@
3.	Smith, L., et al., Molecular dynamics study of deformation and fracture in a tantalum nano-crystalline thin film. Modelling and Simulation in Materials Science and Engineering, 2014. 22(4): p. 045010.
4.	Farkas, D., Fracture Resistance of Nanocrystalline Ni. Metallurgical and Materials Transactions A, 2007. 38(13): p. 2168-2173.
5.	Ruda, M., D. Farkas, and G. Bertolino, Twinning and phase transformations in Zr crack tips. Computational Materials Science, 2010. 49(4): p. 743-750.
6.	Farkas, D., Atomistic simulations of metallic microstructures. Current Opinion in Solid State and Materials Science, 2013. 17(6): p. 284-297.
```
this file is mistformatted and doesn't belong into a source folder.

akohlmey commented on src/USER-SICRACK/README.txt:
```diff
@@ -0,0 +1,11 @@
+How to add a fix to LAMMPS:
```
the instructions here are WAAAY outdated and incorrect and not really needed.

akohlmey commented on src/USER-SICRACK/README.txt:
```diff
@@ -0,0 +1,11 @@
+
+4. Edit "Makefile" in the src-directory, add "user-sicrack" in the list of package variables among the other alternative packages, in the "PACKUSER =" command
+
+5. Make LAMMPS again.
```
also, please no pictures in source code folders.

akohlmey commented on src/USER-SICRACK/fix_sicrack.cpp:
```diff
@@ -0,0 +1,318 @@
+   //Mode 1 crack
+/*------------------------------------------------------------------------- */
+
+#include "math.h"
```
this file violates several conventions for how we format and order include statements. please see the corresponding discussion of the policy in https://github.com/lammps/lammps/blob/master/.github/CONTRIBUTING.md#contributing-code and https://github.com/lammps/lammps/blob/master/doc/include-file-conventions.md

akohlmey commented on src/USER-SICRACK/fix_sicrack.cpp:
```diff
@@ -0,0 +1,318 @@
+
+  // read arguments from input and check validity
+
+  xtip = atof(arg[3]);  // Give coords in (rotated) lattice units, not in Ångstrøm
```
please use `utils::numeric()` instead of `atof()`

akohlmey commented on src/USER-SICRACK/fix_sicrack.cpp:
```diff
@@ -0,0 +1,318 @@
+  int iarg = 7;
+
+  while (iarg < narg) {
+    if (strcmp(arg[iarg],"M1") == 0) {
```
keywords should be all lower case

akohlmey commented on src/USER-SICRACK/fix_sicrack.cpp:
```diff
@@ -0,0 +1,318 @@
+int FixSICrack::setmask()
+{
+  int mask = 0;
+  mask |= LAMMPS_NS::FixConst::INITIAL_INTEGRATE;
```
no need to use the `LAMMPS_NS::` prefix here since you already have a `using LAMMPS_NS;`

akohlmey commented on src/USER-SICRACK/fix_sicrack.cpp:
```diff
@@ -0,0 +1,318 @@
+	cost = cos(theta);
+	sint = sin(theta);
+
+	ux = 0.5/mu*sqrt(0.5*rdist/PI)*(dK1*cost*(Xk - 1.0 + 2.0*sint*sint) + dK2*sint*(Xk + 1.0 + 2.0*cost*cost))*100;
```
please use MY_PI from `math_const.h` for consistency.

akohlmey commented on src/USER-SICRACK/fix_sicrack.h:
```diff
@@ -0,0 +1,57 @@
+#ifndef LMP_FIX_SICRACK_H
+#define LMP_FIX_SICRACK_H
+
+#include "stdio.h"
```
no need to include stdio.h here

akohlmey commented on src/USER-SICRACK/in.sicrackfoam_M1:
```diff
@@ -0,0 +1,62 @@
+log		log.sicrackfoam  # Name of log file to write to
```
no input examples in the source folders please.

bnathan2 commented:
Hello Axel,

Thank you for your response!  This is my first rodeo with both Git and
adapting a code for LAMMPS, and I appreciate the rapid reply and comments.

I have some questions before I can proceed with modifications.

1. Where should I put example scripts?  Should they be in "bench" or in
their own separate directory in the lammps parent directory?

2. With regard to the documentation file, how should it be formatted?
Should it be a PDF?  Should this go in the same folder with the input
scripts?

I've accepted your invitation, so you can modify at will.  As soon as I
hear from you, I'll start working on it.  Thank you again.

Kind Regards.
Nathan



On Thu, Aug 15, 2019 at 7:23 PM Axel Kohlmeyer <notifications@github.com>
wrote:

> *@akohlmey* requested changes on this pull request.
>
> since this is a single style, it should be added to the USER-MISC package
> instead of making it a separate package. it also needs to be better
> included into the documentation. examples need to be small and usable
> directly from the respective folder and moved to examples/USER/misc/sicrack
> ------------------------------
>
> In bench/in.sicrackfoam_M1
> <https://github.com/lammps/lammps/pull/1635#discussion_r314533083>:
>
> > +neighbor    2.0 bin
>
> +neigh_modify    delay 5
>
> +
>
> +## create geometry
>
> +
>
> +read_data CrackBlock
>
> +
>
> +# define groups
>
> +
>
> +region	    mobile block -70.0 70.0 -70.0 70.0 INF INF
>
> +group		mobile region mobile
>
> +group		boundary subtract all mobile
>
> +#set		    group boundary type 2
>
> +
>
> +pair_style eam
>
> +pair_coeff * * /Users/bnathan2/Desktop/LAMMPS_git_Aug14_2019/potentials/Au_u3.eam
>
>
> please don't put absolute paths in input examples, and definitely not
> paths that won't work. also the data file is WAAAY too large for inclusion
> in the distribution and it should not be added into the bench folder.
> please also put a log command into example inputs.
> ------------------------------
>
> In src/USER-SICRACK/Install.csh
> <https://github.com/lammps/lammps/pull/1635#discussion_r314533228>:
>
> > @@ -0,0 +1,13 @@
>
> +# Install/unInstall package "USER-SICRACK" classes in LAMMPS
>
>
> we use bash script only. no c-shell.
> ------------------------------
>
> In src/USER-SICRACK/Package_SiCrack_Documentation.txt
> <https://github.com/lammps/lammps/pull/1635#discussion_r314533362>:
>
> > @@ -0,0 +1 @@
>
> +USER-SICRACK package
>
>
>
> Contents:
>
>
>
> SICRACK stands for semi-infinite crack. This package implements in LAMMPS a fix sicrack command to introduce a semi-infinite crack in a simulation block. The fix enables the user to introduce a semi-infinite crack with desired tip intensity, position, and mode. This is done by applying the equations of linear elastic fracture mechanics for displacement to all atoms in the block. A �skin� of atoms around the border parallel to the crack front should be kept fixed whereas the atoms around the crack tip are free to deform according to the formalism of molecular dynamics and the specified potential.  The package requires no external libraries, and multiple fracture modes and intensities can be combined based on the needs of the user.
>
>
>
> Authors: Nathan Beets and Diana Farkas (Virginia Tech)
>
>
>
> Install:
>
>
>
> This package is installed using the standard instructions for packages in LAMMPS.  One myst be sure to install the �Manybody� potential for this to work properly.  It does not require any specific installation instructions.
>
>
>
> Supporting information:
>
>
>
>
>
>  Syntax of fix sicrack:
>
>
>
> Below is the structure of the fix as coded with this package:
>
>
>
> Fix ID Group-ID sicrack Tip_X Tip_Y G ? Mode_val K_initial K_increase
>
>
>
> ID, Group-ID are documented in the LAMMPS manual
>
> Sicrack = style name of this fix command.
>
> Tip_X = The X position of the crack tip, and the direction in which the crack will advance.
>
> Tip_Y = the Y position of the crack tip.
>
> G = Shear modulus of the material in question (GPa)
>
> ? = the Poisson Ratio of the material in question.
>
> Mode_Val = Crack mode (M1, M2,M3)
>
> K_initial = Applied stress intensity factor to open the first K value to
>
> K_increase = rate (per step) to increase the applied stress intensity by MD step.
>
>
>
> Note: Multiple fixes for multiple modes can be included in the same script and can be run simultaneously. The crack front will always be along the Z axis. The package is designed to run using units �metal�. The shear modulus should be entered in GPa, and the stress intensity in MPa�m.
>
>
>
>  Examples:
>
>
>
> fix     1 all sicrack 200.  200.  27.0 0.44 M1 1.21 0.0
>
> fix     1 all sicrack 200.  200.  27.0 0.44 M2 0.0 0.00001
>
>
>
>  Description:
>
>
>
> This command allows one to nucleate a crack in an atom block with a specific applied stress intensity magnitude K,  in MPa�m and then open the crack by  increasing the K value, according to the displacement equations of linear elastic fracture mechanics, given below [1]:
>
>
>
> (See User_Sicrack_equations)
>
>
>
> Wherein   KI, II, III  are all tip stress intensity values, with units MPa�m, ? is the shear modulus, and ? is the Poisson ratio of your material. ? and r are the angle and distance to the crack tip in the x-y plane, respectively.  The program introduces a crack in the sample with a user-specified crack tip intensity value (K), with the mode specified.  Multiple different fixes can be defined in the same script if a user wants to propagate a crack with more than just one mode.  Once introduced, the crack can be expanded at a user-defined rate, with the K_increase parameter, per MD step.
>
>
>
> It is important to note that if modes II or III are used without an additional fix to include a mode I contribution, any displaced atoms on the crack front that are close enough to the other side will re-bond, eliminating the defined front.  This is specifically different from the method of introducing a crack defined on the LAMMPS website because for this method, no atoms are deleted � only moved according to the equations given above.  So, in order to properly study mode II or III fracture behavior, a small mode I contribution must be added, to allow some separation between atoms on the crack front to deter bond formation.  Otherwise, the user must eliminate atoms along the crack front using the centrosymmetry or common neighbor analysis algorithms.
>
>
>
> This fix can be used for different crack modes (I, II, and III) such that behavior at and around the crack tip is governed by the molecular dynamics formalism, while the sample border is governed by the continuum fracture mechanics equations given above. In order to do this, the best practice is to first define a �border� or skin of atoms around the sample. To do this, utilize the �region� and �group� commands in tandem with the samples� dimensions (see example scripts). This border needs to be somewhat larger than the range of the interatomic potential used and the atomic positions are kept fixed.  As an example, if you have a 40 nm by 40 nm by 40 nm cuboid structure filled with a single crystal of gold (lattice parameter 0.408 nm) with origin in a corner and all initial atom coordinates positive, the crack front (as specified by the fix) will be parallel to the z axis, and the crack itself will oriented in the x direction.  The skin should be around 2.0 nm thick, parallel to the crack front.  So, this is all atoms where x > 38, x < 2, and y > 38, y < 2.  All atoms in this region will be defined as group �border.�  Everything outside of it will be defined as group �mobile.� An example of this is documented in either of the attached input scripts.
>
>
>
> Figure 1 illustrates this for a mode 1 crack.
>
>
>
> (See User_Sicrack_Figure1)
>
>
>
> Figure 1: Border or �fixed� atoms and nucleation of a mode I crack.  Similar setups can be used for modesII and III.
>
>
>
> After a border has been defined, the user needs to open the crack to a specified intensity, through the fix, which is applied to all atoms (fixed and free, from the figure above).  Ideally, this initial opening should be at a value for critical stress intensity so that the crack will not close immediately after taking the formalism away.  Determining this optimum value for K usually takes a little trial and error.  For typical metals, a stress intensity of around K = 1.2 MPa?m is a good place to start.  In order to study crack propagation, this first fix command can be unfixed and a second fix can be applied with K_initial as zero, but K_increase is used to introduce additional crack opening.  This value is the rate of increase of K per step.  The motion of the interior free atoms is governed by the molecular dynamics, while atom motion in the border, far from the crack tip is given by linear elastic fracture mechanics.
>
>
>
> This formalism has been used for to simulate fracture behavior in zirconium, [2] tantalum [3] and nickel [4] . It has also been used to model microstructural evolution in the crack tip region in single and polycrystalline systems. [5, 6]
>
>
>
>
>
> Restrictions:
>
>
>
> The use of this fix is limited to the nvt and nve ensembles.  Furthermore, due to the fact that the samples sides perpendicular to the Z direction are being continuously deformed, we cannot have periodic boundary conditions in either the x or y directions.  Both periodic or free surface boundary conditions along the Z axis are possible.   Finally, if simulating a brittle material, where the crack tip is expected to propagate into the sample with increasing applied stress tip intensity, the user will need to manually move the crack tip forward by altering the input script and keeping up with the tip by checking the dump files, or sample snapshots as the deformation proceeds.
>
>
>
> Related Commands:
>
>
>
> None
>
>
>
> Default:
>
>
>
> None
>
>
>
>
>
> Bibliography:	
>
>
>
> 1.	Liebowitz, H. and G. C. Sih, Mathematical Theories of Brittle Fracture. Vol. Vol. II. 1968.
>
> 2.	Bertolino, G., M. Ruda, and D. Farkas, Fracture resistance of textured polycrystalline Zr: A simulation study. Computational Materials Science, 2019. 162: p. 304-313.
>
> 3.	Smith, L., et al., Molecular dynamics study of deformation and fracture in a tantalum nano-crystalline thin film. Modelling and Simulation in Materials Science and Engineering, 2014. 22(4): p. 045010.
>
> 4.	Farkas, D., Fracture Resistance of Nanocrystalline Ni. Metallurgical and Materials Transactions A, 2007. 38(13): p. 2168-2173.
>
> 5.	Ruda, M., D. Farkas, and G. Bertolino, Twinning and phase transformations in Zr crack tips. Computational Materials Science, 2010. 49(4): p. 743-750.
>
> 6.	Farkas, D., Atomistic simulations of metallic microstructures. Current Opinion in Solid State and Materials Science, 2013. 17(6): p. 284-297.
>
>
>
>
>
>
> this file is mistformatted and doesn't belong into a source folder.
> ------------------------------
>
> In src/USER-SICRACK/README.txt
> <https://github.com/lammps/lammps/pull/1635#discussion_r314533494>:
>
> > @@ -0,0 +1,11 @@
>
> +How to add a fix to LAMMPS:
>
>
> the instructions here are WAAAY outdated and incorrect and not really
> needed.
> ------------------------------
>
> In src/USER-SICRACK/README.txt
> <https://github.com/lammps/lammps/pull/1635#discussion_r314533621>:
>
> > @@ -0,0 +1,11 @@
>
> +How to add a fix to LAMMPS:
>
> +
>
> +1. Put the .h and .cpp files in the src/ folder, and make sure it follows the same style as other fix files already included in LAMMPS.
>
> +
>
> +2. Edit style_fix.h and add an your fix_foo.h file to the list of other fixes.
>
> +
>
> +3. Put fix_sicrack files and the other crack script files in separate folder "USER-SICRACK", and make an "Install.csh" script which will (re)move files to LAMMPS src directory when the command "make yes-user-sicrack" or "make no-user-sicrack" is issued before making LAMMPS
>
> +
>
> +4. Edit "Makefile" in the src-directory, add "user-sicrack" in the list of package variables among the other alternative packages, in the "PACKUSER =" command
>
> +
>
> +5. Make LAMMPS again.
>
>
> also, please no pictures in source code folders.
> ------------------------------
>
> In src/USER-SICRACK/fix_sicrack.cpp
> <https://github.com/lammps/lammps/pull/1635#discussion_r314534181>:
>
> > +/* ----------------------------------------------------------------------
>
> +   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
>
> +   www.cs.sandia.gov/~sjplimp/lammps.html
>
> +   Steve Plimpton, sjplimp@sandia.gov, Sandia National Laboratories
>
> +
>
> +   Copyright (2003) Sandia Corporation.  Under the terms of Contract
>
> +   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
>
> +   certain rights in this software.  This software is distributed under
>
> +   the GNU General Public License.
>
> +
>
> +   See the README file in the top-level LAMMPS directory.
>
> +   //***CETTINA
>
> +   //Mode 1 crack
>
> +/*------------------------------------------------------------------------- */
>
> +
>
> +#include "math.h"
>
>
> this file violates several conventions for how we format and order include
> statements. please see the corresponding discussion of the policy in
> https://github.com/lammps/lammps/blob/master/.github/CONTRIBUTING.md#contributing-code
> and
> https://github.com/lammps/lammps/blob/master/doc/include-file-conventions.md
> ------------------------------
>
> In src/USER-SICRACK/fix_sicrack.cpp
> <https://github.com/lammps/lammps/pull/1635#discussion_r314534362>:
>
> > +  Fix(lmp, narg, arg)
>
> +{
>
> +  if (narg < 10) error->all(FLERR,"Illegal fix sicrack command");
>
> +
>
> +  K1 = 0.0;
>
> +  dK1 = 0.0;
>
> +  K2 = 0.0;
>
> +  dK2 = 0.0;
>
> +  K3 = 0.0;
>
> +  dK3 = 0.0;
>
> +  T = 0.0;
>
> +  dT = 0.0;
>
> +
>
> +  // read arguments from input and check validity
>
> +
>
> +  xtip = atof(arg[3]);  // Give coords in (rotated) lattice units, not in Ångstrøm
>
>
> please use utils::numeric() instead of atof()
> ------------------------------
>
> In src/USER-SICRACK/fix_sicrack.cpp
> <https://github.com/lammps/lammps/pull/1635#discussion_r314534510>:
>
> > +  K3 = 0.0;
>
> +  dK3 = 0.0;
>
> +  T = 0.0;
>
> +  dT = 0.0;
>
> +
>
> +  // read arguments from input and check validity
>
> +
>
> +  xtip = atof(arg[3]);  // Give coords in (rotated) lattice units, not in Ångstrøm
>
> +  ytip = atof(arg[4]);
>
> +  mu = atof(arg[5]);   // Shear modulus, in GPa
>
> +  vu = atof(arg[6]);  // Poisson ratio
>
> +
>
> +  int iarg = 7;
>
> +
>
> +  while (iarg < narg) {
>
> +    if (strcmp(arg[iarg],"M1") == 0) {
>
>
> keywords should be all lower case
> ------------------------------
>
> In src/USER-SICRACK/fix_sicrack.cpp
> <https://github.com/lammps/lammps/pull/1635#discussion_r314534779>:
>
> > +
>
> +  atom->delete_callback(id,0);
>
> +  atom->delete_callback(id,1);
>
> +
>
> +  // delete locally stored arrays
>
> +
>
> +  memory->destroy(xoriginal);
>
> +}
>
> +
>
> +/* ---------------------------------------------------------------------- */
>
> +
>
> +
>
> +int FixSICrack::setmask()
>
> +{
>
> +  int mask = 0;
>
> +  mask |= LAMMPS_NS::FixConst::INITIAL_INTEGRATE;
>
>
> no need to use the LAMMPS_NS:: prefix here since you already have a using
> LAMMPS_NS;
> ------------------------------
>
> In src/USER-SICRACK/fix_sicrack.cpp
> <https://github.com/lammps/lammps/pull/1635#discussion_r314535005>:
>
> > +      xold[2] = x[i][2];
>
> +
>
> +      xtemp = x[i][0] - xtip;
>
> +      ytemp = x[i][1] - ytip;
>
> +      rdist = sqrt(xtemp*xtemp + ytemp*ytemp);
>
> +
>
> +      // theta is the angle made by vector (x,y) with the crack plane
>
> +
>
> +      // a domain error occurs if both xtemp and ytemp are 0
>
> +      if (rdist != 0) {
>
> +      	theta2 = atan2(ytemp,xtemp);
>
> +	theta = 0.5*theta2;
>
> +	cost = cos(theta);
>
> +	sint = sin(theta);
>
> +
>
> +	ux = 0.5/mu*sqrt(0.5*rdist/PI)*(dK1*cost*(Xk - 1.0 + 2.0*sint*sint) + dK2*sint*(Xk + 1.0 + 2.0*cost*cost))*100;
>
>
> please use MY_PI from math_const.h for consistency.
> ------------------------------
>
> In src/USER-SICRACK/fix_sicrack.h
> <https://github.com/lammps/lammps/pull/1635#discussion_r314535130>:
>
> > +
>
> +   See the README file in the top-level LAMMPS directory.
>
> +
>
> +   Mode I, II or III semi-infinite crack, with T-stress
>
> +/*------------------------------------------------------------------------- */
>
> +
>
> +#ifdef FIX_CLASS
>
> +
>
> +FixStyle(sicrack,FixSICrack)
>
> +
>
> +#else
>
> +
>
> +#ifndef LMP_FIX_SICRACK_H
>
> +#define LMP_FIX_SICRACK_H
>
> +
>
> +#include "stdio.h"
>
>
> no need to include stdio.h here
> ------------------------------
>
> In src/USER-SICRACK/in.sicrackfoam_M1
> <https://github.com/lammps/lammps/pull/1635#discussion_r314535264>:
>
> > @@ -0,0 +1,62 @@
>
> +log		log.sicrackfoam  # Name of log file to write to
>
>
> no input examples in the source folders please.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/1635?email_source=notifications&email_token=AL3TW3FZL2PM7L6MPYXXSJDQEXQN3A5CNFSM4IMBGFNKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBXQXGA#pullrequestreview-275712920>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3TW3AKHSLMCXAPYSHWMCTQEXQN3ANCNFSM4IMBGFNA>
> .
>


-- 
Nathan Beets

PhD Student, Virginia Tech Department of Materials Science and Engineering

akohlmey commented:
> Hello Axel, Thank you for your response! This is my first rodeo with both Git and adapting a code for LAMMPS, 

no problem, we all have to start somewhere. we are quite used to folks being first timers. ;-)

>and I appreciate the rapid reply and comments. I have some questions before I can proceed with modifications. 
> 1. Where should I put example scripts? Should they be in "bench" or in their own separate directory in the lammps parent directory?

as noted in one of my comments. since this fix style should be included in the USER-MISC package, the example should be moved to `examples/USER/misc/sicrack`

> 2. With regard to the documentation file, how should it be formatted? Should it be a PDF?
> Should this go in the same folder with the input scripts?

the reference documentation of the fix should go into the doc/src/ folder and has to follow the usual txt2html markup, that LAMMPS uses. you have this file already, you just need to move it and add under restrictions the usual disclaimer, that this fix is part of the USER-MISC package. the remainder of your text should be in a properly wrapped text file alongside the examples and written mostly in the style of a how-to (which is also mostly done). you should remove all obsolete and incorrect comments, e.g. about installation and how to merge the "package" with LAMMPS.

> I've accepted your invitation, so you can modify at will. As soon as I hear from you, I'll start working on it. Thank you again. Kind Regards. Nathan

please be aware, that this was a first quick pass at reviewing your contribution. there will likely be more comments and checks. we usually look and comment on simple formal issues first and will have a more in depth look after everything fits in cleanly with the rest of the software.

> 
> -- Nathan Beets PhD Student, Virginia Tech Department of Materials Science and Engineering

bnathan2 commented:
Superb! I'll get working on it, and push a new version after I have
incorporated your suggestions.  Thanks again.

On Fri, Aug 16, 2019 at 9:54 AM Axel Kohlmeyer <notifications@github.com>
wrote:

> Hello Axel, Thank you for your response! This is my first rodeo with both
> Git and adapting a code for LAMMPS,
>
> no problem, we all have to start somewhere. we are quite used to folks
> being first timers. ;-)
>
> and I appreciate the rapid reply and comments. I have some questions
> before I can proceed with modifications.
>
>    1. Where should I put example scripts? Should they be in "bench" or in
>    their own separate directory in the lammps parent directory?
>
> as noted in one of my comments. since this fix style should be included in
> the USER-MISC package, the example should be moved to
> examples/USER/misc/sicrack
>
>
>    1. With regard to the documentation file, how should it be formatted?
>    Should it be a PDF?
>    Should this go in the same folder with the input scripts?
>
> the reference documentation of the fix should go into the doc/src/ folder
> and has to follow the usual txt2html markup, that LAMMPS uses. you have
> this file already, you just need to move it and add under restrictions the
> usual disclaimer, that this fix is part of the USER-MISC package. the
> remainder of your text should be in a properly wrapped text file alongside
> the examples and written mostly in the style of a how-to (which is also
> mostly done). you should remove all obsolete and incorrect comments, e.g.
> about installation and how to merge the "package" with LAMMPS.
>
> I've accepted your invitation, so you can modify at will. As soon as I
> hear from you, I'll start working on it. Thank you again. Kind Regards.
> Nathan
>
> please be aware, that this was a first quick pass at reviewing your
> contribution. there will likely be more comments and checks. we usually
> look and comment on simple formal issues first and will have a more in
> depth look after everything fits in cleanly with the rest of the software.
>
> -- Nathan Beets PhD Student, Virginia Tech Department of Materials Science
> and Engineering
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/1635?email_source=notifications&email_token=AL3TW3CZ5ZTFBGBJIFUK4BDQE2WRDA5CNFSM4IMBGFNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4OVRUQ#issuecomment-522016978>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3TW3CJJ4U44H7VBBKFZU3QE2WRDANCNFSM4IMBGFNA>
> .
>


-- 
Nathan Beets

PhD Student, Virginia Tech Department of Materials Science and Engineering

bnathan2 commented:
Dear Axel,

I have pushed an updated version of lammps to my user page,
bnathan2/lammps.  Directories with changs have been marked
"August21version"  If you need me to make another pull request for you to
view and comment on the changes, please let me know.  The only part of your
comments I couldn't figure out was where I needed to replace "atof" with
utils::numeric().  when I ran with utils::numeric(), an error message told
me I had too few arguments, despitr trying several different
configurations. The code still functions as desired, but if you need it
changed, I'd be very grateful if you could point me in the direction of a
"utils" doc page.

Attached is a file with responses and simple descriptions of what I
changed.  Let me know what you need when possible.  Thank you again for
your help. :)

Best,
Nathan



On Fri, Aug 16, 2019 at 10:17 AM Nathan Beets <bnathan2@vt.edu> wrote:

>
>
> ---------- Forwarded message ---------
> From: Nathan Beets <bnathan2@vt.edu>
> Date: Fri, Aug 16, 2019 at 10:13 AM
> Subject: Re: [lammps/lammps] New user package for semi infinite crack
> propagation (#1635)
> To: lammps/lammps <
> reply@reply.github.com>
>
>
> Superb! I'll get working on it, and push a new version after I have
> incorporated your suggestions.  Thanks again.
>
> On Fri, Aug 16, 2019 at 9:54 AM Axel Kohlmeyer <notifications@github.com>
> wrote:
>
>> Hello Axel, Thank you for your response! This is my first rodeo with both
>> Git and adapting a code for LAMMPS,
>>
>> no problem, we all have to start somewhere. we are quite used to folks
>> being first timers. ;-)
>>
>> and I appreciate the rapid reply and comments. I have some questions
>> before I can proceed with modifications.
>>
>>    1. Where should I put example scripts? Should they be in "bench" or
>>    in their own separate directory in the lammps parent directory?
>>
>> as noted in one of my comments. since this fix style should be included
>> in the USER-MISC package, the example should be moved to
>> examples/USER/misc/sicrack
>>
>>
>>    1. With regard to the documentation file, how should it be formatted?
>>    Should it be a PDF?
>>    Should this go in the same folder with the input scripts?
>>
>> the reference documentation of the fix should go into the doc/src/ folder
>> and has to follow the usual txt2html markup, that LAMMPS uses. you have
>> this file already, you just need to move it and add under restrictions the
>> usual disclaimer, that this fix is part of the USER-MISC package. the
>> remainder of your text should be in a properly wrapped text file alongside
>> the examples and written mostly in the style of a how-to (which is also
>> mostly done). you should remove all obsolete and incorrect comments, e.g.
>> about installation and how to merge the "package" with LAMMPS.
>>
>> I've accepted your invitation, so you can modify at will. As soon as I
>> hear from you, I'll start working on it. Thank you again. Kind Regards.
>> Nathan
>>
>> please be aware, that this was a first quick pass at reviewing your
>> contribution. there will likely be more comments and checks. we usually
>> look and comment on simple formal issues first and will have a more in
>> depth look after everything fits in cleanly with the rest of the software.
>>
>> -- Nathan Beets PhD Student, Virginia Tech Department of Materials
>> Science and Engineering
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/lammps/lammps/pull/1635?email_source=notifications&email_token=AL3TW3CZ5ZTFBGBJIFUK4BDQE2WRDA5CNFSM4IMBGFNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4OVRUQ#issuecomment-522016978>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AL3TW3CJJ4U44H7VBBKFZU3QE2WRDANCNFSM4IMBGFNA>
>> .
>>
>
>
> --
> Nathan Beets
>
> PhD Student, Virginia Tech Department of Materials Science and Engineering
>
>
> --
> Nathan Beets
>
> PhD Student, Virginia Tech Department of Materials Science and Engineering
>


-- 
Nathan Beets

PhD Student, Virginia Tech Department of Materials Science and Engineering

akohlmey commented:
> Dear Axel, I have pushed an updated version of lammps to my user page, bnathan2/lammps. Directories with changs have been marked "August21version" If you need me to make another pull request for you to view and comment on the changes, please let me know. 

there are no changes visible in this branch that you submitted as pull request.

> The only part of your comments I couldn't figure out was where I needed to replace "atof" with utils::numeric(). when I ran with utils::numeric(), an error message told me I had too few arguments, 

yes, the calling sequence is slightly different, but all you need to do is to look at existing usage in other parts of LAMMPS. for the most part, you have to insert an FLERR tag, which expands to the line number and file name, hence the warning about missing arguments.

> despitr trying several different configurations. The code still functions as desired, but if you need it changed, I'd be very grateful if you could point me in the direction of a "utils" doc page. Attached is a file with responses and simple descriptions of what I changed. Let me know what you need when possible. Thank you again for your help. :) Best, Nathan

attachments get scrubbed when things pass through github. so i cannot see that either.
it is generally better to converse through using the web interface instead of using e-mails (or stick to plain text with no attachments).

axel.


> […](#)
> On Fri, Aug 16, 2019 at 10:17 AM Nathan Beets ***@***.***> wrote: ---------- Forwarded message --------- From: Nathan Beets ***@***.***> Date: Fri, Aug 16, 2019 at 10:13 AM Subject: Re: [lammps/lammps] New user package for semi infinite crack propagation (#1635) To: lammps/lammps < ***@***.***> Superb! I'll get working on it, and push a new version after I have incorporated your suggestions. Thanks again. On Fri, Aug 16, 2019 at 9:54 AM Axel Kohlmeyer ***@***.***> wrote: > Hello Axel, Thank you for your response! This is my first rodeo with both > Git and adapting a code for LAMMPS, > > no problem, we all have to start somewhere. we are quite used to folks > being first timers. ;-) > > and I appreciate the rapid reply and comments. I have some questions > before I can proceed with modifications. > > 1. Where should I put example scripts? Should they be in "bench" or > in their own separate directory in the lammps parent directory? > > as noted in one of my comments. since this fix style should be included > in the USER-MISC package, the example should be moved to > examples/USER/misc/sicrack > > > 1. With regard to the documentation file, how should it be formatted? > Should it be a PDF? > Should this go in the same folder with the input scripts? > > the reference documentation of the fix should go into the doc/src/ folder > and has to follow the usual txt2html markup, that LAMMPS uses. you have > this file already, you just need to move it and add under restrictions the > usual disclaimer, that this fix is part of the USER-MISC package. the > remainder of your text should be in a properly wrapped text file alongside > the examples and written mostly in the style of a how-to (which is also > mostly done). you should remove all obsolete and incorrect comments, e.g. > about installation and how to merge the "package" with LAMMPS. > > I've accepted your invitation, so you can modify at will. As soon as I > hear from you, I'll start working on it. Thank you again. Kind Regards. > Nathan > > please be aware, that this was a first quick pass at reviewing your > contribution. there will likely be more comments and checks. we usually > look and comment on simple formal issues first and will have a more in > depth look after everything fits in cleanly with the rest of the software. > > -- Nathan Beets PhD Student, Virginia Tech Department of Materials > Science and Engineering > > — > You are receiving this because you authored the thread. > Reply to this email directly, view it on GitHub > <#1635?email_source=notifications&email_token=AL3TW3CZ5ZTFBGBJIFUK4BDQE2WRDA5CNFSM4IMBGFNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4OVRUQ#issuecomment-522016978>, > or mute the thread > <https://github.com/notifications/unsubscribe-auth/AL3TW3CJJ4U44H7VBBKFZU3QE2WRDANCNFSM4IMBGFNA> > . > -- Nathan Beets PhD Student, Virginia Tech Department of Materials Science and Engineering -- Nathan Beets PhD Student, Virginia Tech Department of Materials Science and Engineering
> -- Nathan Beets PhD Student, Virginia Tech Department of Materials Science and Engineering

bnathan2 commented:
Working on it.  Trying to merge the modified version into the submitted version. 

whenever I try to clone from or push to 
https://github.com/bnathan2/lammps/tree/NewUserPackage_SemiInfiniteCrack_NathanBeets

that the address is not found. Should I be pushing/cloning from a different address?

akohlmey commented:
> Working on it. Trying to merge the modified version into the submitted version.
> 
> whenever I try to clone from or push to
> https://github.com/bnathan2/lammps/tree/NewUserPackage_SemiInfiniteCrack_NathanBeets
> 
> that the address is not found. Should I be pushing/cloning from a different address?

yes. please have a look at: https://lammps.sandia.gov/doc/Howto_github.html

bnathan2 commented:
Alright! I was able to re-write the section of the .cpp file using utils::numeric() instead of atof(), and have pushed an updated repository into this branch.  Sorry about the wait. Looking forward to your comments. Response document attached. 

[Axel Comments and responses August 21.txt](https://github.com/lammps/lammps/files/3537827/Axel.Comments.and.responses.August.21.txt)

akohlmey commented:
> Alright! I was able to re-write the section of the .cpp file using utils::numeric() instead of atof(), and have pushed an updated repository into this branch. Sorry about the wait. Looking forward to your comments. Response document attached.
> 
> [Axel Comments and responses August 21.txt](https://github.com/lammps/lammps/files/3537827/Axel.Comments.and.responses.August.21.txt)

sorry, but this pull request has become a big mess as it tries to change the permissions on all files so it is extremely laborious to figure out the changes. There are now also loads of merge conflicts. I suggest to start over making a new branch from an up-to-date master and try with a new pull request based on that branch with only the minimal changes and new files in the proper locations.

bnathan2 commented:
Ok, I'll cancel this one and start a new one.
